aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,329 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+// Iterates over first `n_batck_dims` and applies `f` to each element.
+// TODO: this implemtation is not efficient, it simply iteratates over
+// batch elements in a serial for-loop.
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes;
+    tensor_view_sizes.push_back(-1);
+    for (int d = n_batch_dims; d < tensor.dim(); ++d) {
+      tensor_view_sizes.push_back(tensor.size(d));
+    }
+    auto tensor_view = tensor.view(tensor_view_sizes);
+    tensors_view.push_back(tensor_view);
+  }
+
+  for (int64_t input_idx = 0; input_idx < tensors_view[0].size(0);
+    ++input_idx) {
+    std::vector<Tensor> input_selected;
+    for (auto& tensor_view : tensors_view) {
+      auto curr_input = tensor_view.select(0, input_idx);
+      input_selected.push_back(curr_input);
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
+  if (matrices.dim() > 2) {
+    auto res = at::empty(matrices.sizes(), matrices.options());
+
+    batch_apply(
+      {res, matrices, powers},
+      matrices.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+        int64_t n = tensors[2].item().template to<int64_t>();
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + A2 / 2;
+}
+
+Tensor compute_T4(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / 2. + A / 6. + A2 / 24.);
+}
+
+template <typename scalar_t>
+Tensor compute_T8(const Tensor& A) {
+  // no constexpr support for std::sqrt unfortunately
+  scalar_t sqrt_177 = std::sqrt(177);
```

Comment:
nit: 177.
It might be good to put a dot at the end of the numbers to indicate that they are floating point. This should not be a problem in the expressions below, as they all involve the floating point `sqrt_177`, but still